src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -586,55 +736,76 @@ export class SavedObjectsRepository {
       return { saved_objects: [] };
     }
 
-    const unsupportedTypeObjects = objects
-      .filter(o => !this._allowedTypes.includes(o.type))
-      .map(({ type, id }) => {
-        return ({
-          id,
-          type,
-          error: SavedObjectsErrorHelpers.createUnsupportedTypeError(type).output.payload,
-        } as any) as SavedObject<T>;
-      });
+    let bulkGetRequestIndexCounter = 0;
+    const bulkGetDocs: object[] = [];
+    const expectedBulkGetResults: Array<Either<any, any>> = objects.map(object => {
+      const { type, id, fields } = object;
+
+      if (!this._allowedTypes.includes(type)) {
+        return {
+          tag: 'Left' as 'Left',
+          error: {
+            id,
+            type,
+            error: SavedObjectsErrorHelpers.createUnsupportedTypeError(type).output.payload,
+          },
+        };
+      }
 
-    const supportedTypeObjects = objects.filter(o => this._allowedTypes.includes(o.type));
+      bulkGetDocs.push({
+        _id: this._serializer.generateRawId(namespace, type, id),
+        _index: this.getIndexForType(type),
+        _source: includedFields(type, fields),
+      });
```

Comment:
If we include `fields` in the right value we replace this with a filter + map:
```
const bulkGetDocs  = expectedBulkGetResults
  .filter(isRight)
  .map(({value: object}) => {{
       const {type, id, fields} = object;
        _id: this._serializer.generateRawId(namespace, type, id),
        _index: this.getIndexForType(type),
        _source: includedFields(type, fields),
      }})
```

This is quite a complex method and I think this might make the code a bit more readable, but it's a matter of taste/opinion, so I'll leave it up to you to decide to include it or not.

Reply:
Makes sense, I had to refactor the `Either` type in a832503 first, and then I was able to make this change in 94d31ec.